x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

```diff
@@ -150,6 +167,99 @@ public static void switchUserBasedOnActionOriginAndExecute(
         }
     }
 
+    public static void maybePreAuthorizeChildAction(
+        ThreadContext threadContext,
+        ClusterState clusterState,
+        String childAction,
+        DiscoveryNode destinationNode,
+        TransportRequest request
+    ) {
+
+        AuthorizationEngine.AuthorizationContext parentContext = extractAuthorizationContext(threadContext, childAction);
+        if (parentContext == null) {
+            return;
+        }
+
+        if (childAction.startsWith(parentContext.getAction()) == false) {
+            // Parent action is not a true parent
+            // We want to treat shard level actions (those that append '[s]' and/or '[p]' & '[r]')
+            // or similar (e.g. search phases) as children, but not every action that is triggered
+            // within another action should be authorized this way
+            return;
+        }
+
+        final IndicesRequest indicesRequest;
+        if (request instanceof IndicesRequest) {
+            indicesRequest = (IndicesRequest) request;
+        } else {
+            // We can only pre-authorize indices request.
+            return;
+        }
+
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0 || Arrays.equals(IndicesAndAliasesResolver.NO_INDICES_OR_ALIASES_ARRAY, indices)) {
+            // No indices to check
+            return;
+        }
+
```

Comment:
Should we also check that the child action refers to a subset of the indices of the parent?